### ISSUE TITLE
Schicht editieren

### ISSUE THREAD

**Issue by sheepyhollow** [mimischi/django-clock#466]
Sollte man laufende (gestochene) Schichten editieren können? Oder geht das?

Angenommen, ich clocke mich 15 Minuten zu spät ein und möchte den Anfang gleich korrigieren (wo ich schonmal dran bin).

Oder ich möchte meine Schicht so einstellen, dass sie sich in 20 Minuten automatisch beendet (weil warum auch immer), obwohl ich sie mit stechen begonnen habe.

**Comment by mimischi** [comment]
> Sollte man laufende (gestochene) Schichten editieren können?
Vielleicht.

> Oder geht das?
Nein.

Behalten wir uns für die Zukunft vor.

**Comment by mimischi** [comment]
Eigentlich dürfte das ganz fix gehen. Welche Einschränkungen wollen wir denn vornehmen, also was sollen die Nutzer hier ändern können?

- Schichtanfang
- Zu gehöriger Vertrag

Mehr ja wahrscheinlich nicht? Wir werden hier allerdings _nicht_ erlauben, dass ein User die Startzeit in die Zukunft setzt. Das heißt:

Schicht gestartet um 08:00, aktuell ist 10:00. User kann die Startzeit nun auf spätestens 09:55 anpassen. Später ist nicht erlaubt, da wir sonst eine negative Schichtdauer hätten.

**Comment by sheepyhollow** [comment]
Genau, die _Vor_datierung des Schichtanfangs sollte möglich sein, falls ich zu spät einclocke. 
Parameter ändern wie Vertrag und ggf. Notizen sollte ja unkritisch sein.
Ich fände nach wie vor das Eingeben eines automatischen Schichtendes sinnvoll, aber da bin ich nicht sicher.

Allerdings muss man sich überlegen, wo man das macht - laufende Schichten stehen ja erst nach dem Ende in der Schichtübersicht, richtig? D.h. entweder müsste man laufende Schichten in die Übersicht bringen oder ich müsste sozusagen neben Start/Stop einen Änderungsbutton haben. Das bläht es natürlich wieder ein bisschen auf...

**Comment by mimischi** [comment]
> Ich fände nach wie vor das Eingeben eines automatischen Schichtendes sinnvoll, aber da bin ich nicht sicher.

Das könnten wir schon einrichten, aber am Ende wäre es ein "Ich gebe jetzt das Schichtende ein und beende damit die aktuell laufende Schicht".

> Allerdings muss man sich überlegen, wo man das macht - laufende Schichten stehen ja erst nach dem Ende in der Schichtübersicht, richtig?

Ich hatte jetzt gedacht, wir fügen irgendwie eine Option (Button) beim Ein-/Ausstechen hinzu. Dieser erscheint nur bei einer laufenden schicht und ermöglicht es, diese zu ändern.

**Comment by sheepyhollow** [comment]
> am Ende wäre es ein "Ich gebe jetzt das Schichtende ein und beende damit die aktuell laufende Schicht"

Hm... ja - das wäre ja auch in Ordnung. Ich würde nur meine Schicht vorzeitig ausstechen können, falls ich nachher nicht dran denke oder alle Hände voll zu tun habe.

> wir fügen irgendwie eine Option (Button) beim Ein-/Ausstechen hinzu. Dieser erscheint nur bei einer laufenden Schicht

Ja, das dachte ich auch - hatte nur überlegt, ob's das nicht überlädt.

**Comment by mimischi** [comment]
> Ja, das dachte ich auch - hatte nur überlegt, ob's das nicht überlädt.

Denke vor erst ist das in Ordnung. Für das überarbeitet UI können wir uns gerne mal mehr Gedanken machen.

**Comment by chgad** [comment]
> > am Ende wäre es ein "Ich gebe jetzt das Schichtende ein und beende damit die aktuell laufende Schicht"
> 
> 
> Hm... ja - das wäre ja auch in Ordnung. Ich würde nur meine Schicht vorzeitig ausstechen können, falls ich nachher nicht dran denke oder alle Hände voll zu tun habe.

Das erachte ich in der Hinsicht als unnötig als es das "ein- und ausclocken" überflüssig macht. Da eine "beende Schicht in Zukunft" einem "ich trage meine Schicht direkt ein" gleicht. Es würde lediglich das System entlasten da so weniger Schichten, die vergessen wurden, laufen würden.

**Comment by mimischi** [comment]
Ich sehe folgenden Fall:

User sticht ein und merkt, dass er eigentlich eine Schicht auch händisch anlegen konnte. Anstatt jetzt die Schicht zu stoppen, und nachträglich zu editieren um die Endzeit anzupassen, kann die Endzeit in der "laufenden Schicht" eingetragen werden.

Für das System ist die Schicht ab dem Zeitpunkt dann auch nicht mehr aktiv sondern beendet. Eine aktive Schicht ist nämlich die, die keine Endzeit hat. Es ist wahrscheinlich eine bessere UX, mehr nicht. Sehe nicht, dass es das ganze System komplexer macht.

**Comment by sheepyhollow** [comment]
Stimme Michael da zu.
Der Unterschied zum direkten Eintragen ist aber, dass der Startpunkt per Clock-Button gesetzt wurde.
Aber - was ist denn der technische Aspekt?

Eigentlich sind die Start-/Stop-Buttons nur sowas wie shortcuts, die `now()` als Start- bzw. Endzeit einsetzen. Ich würde schon erwarten, dass ich beide Wege beschreiten kann - explizite Eintragung oder Shortcut. Aber dann muss eine laufende Schicht eigentlich auch in der Schichtübersicht angezeigt werden. Tatsächlich würde ich das auch erwarten. Da steht dann keine Endzeit sondern "Schicht läuft" oder so. Und dann kann ich die in der Schichtübersicht auch beenden.

**Comment by mimischi** [comment]
> Aber dann muss eine laufende Schicht eigentlich auch in der Schichtübersicht angezeigt werden. Tatsächlich würde ich das auch erwarten. Da steht dann keine Endzeit sondern "Schicht läuft" oder so. Und dann kann ich die in der Schichtübersicht auch beenden.

Das lässt sich ohne Probleme ändern. Aktuell filtern wir einfach alle Schichten heraus, die noch keine Endzeit haben (=laufende Schichten).
